File: androidtv/app/src/main/java/com/nedflix/tv/data/ApiClient.kt
/*
 * Nedflix for Android TV - API Client
 * Handles communication with Nedflix server
 */

package com.nedflix.tv.data

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.Json
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import java.net.URLEncoder
import java.util.concurrent.TimeUnit

class ApiClient private constructor() {
    private var baseUrl: String = ""
    private var authToken: String? = null

    private val json = Json { ignoreUnknownKeys = true }

    private val client = OkHttpClient.Builder()
        .connectTimeout(30, TimeUnit.SECONDS)
        .readTimeout(60, TimeUnit.SECONDS)
        .writeTimeout(30, TimeUnit.SECONDS)
        .build()

    fun configure(serverUrl: String) {
        baseUrl = serverUrl.trimEnd('/')
    }

    fun setToken(token: String?) {
        authToken = token
    }

    // Authentication
    suspend fun authenticate(username: String, password: String): String = withContext(Dispatchers.IO) {
        val requestBody = """{"username":"$username","password":"$password"}"""
            .toRequestBody("application/json".toMediaType())

        val request = Request.Builder()
            .url("$baseUrl/api/auth/login")
            .post(requestBody)
            .build()

        val response = client.newCall(request).execute()
        if (!response.isSuccessful) {
            throw Exception("Authentication failed: ${response.code}")
        }

        val body = response.body?.string() ?: throw Exception("Empty response")
        val authResponse = json.decodeFromString<AuthResponse>(body)
        authToken = authResponse.token

        // Save token
        SettingsManager.instance.updateToken(authResponse.token)

        authResponse.token
    }

    // Fetch library
    suspend fun fetchLibrary(): List<MediaCategory> = withContext(Dispatchers.IO) {
        val requestBuilder = Request.Builder()
            .url("$baseUrl/api/library")
            .get()

        authToken?.let {
            requestBuilder.addHeader("Authorization", "Bearer $it")
        }

        val response = client.newCall(requestBuilder.build()).execute()

        if (response.code == 401) {
            throw Exception("Authentication required")
        }

        if (!response.isSuccessful) {
            throw Exception("Server error: ${response.code}")
        }

        val body = response.body?.string() ?: "[]"

        // Try parsing as categories first
        try {
            return@withContext json.decodeFromString<List<MediaCategory>>(body)
        } catch (e: Exception) {
            // Try parsing as flat list
            try {
                val items = json.decodeFromString<List<MediaItem>>(body)
                return@withContext groupItemsByType(items)
            } catch (e2: Exception) {
                return@withContext emptyList()
            }
        }
    }

    private fun groupItemsByType(items: List<MediaItem>): List<MediaCategory> {
        val videos = mutableListOf<MediaItem>()
        val audio = mutableListOf<MediaItem>()
        val other = mutableListOf<MediaItem>()

        items.forEach { item ->
            when (item.type.lowercase()) {
                "video", "movie", "tv" -> videos.add(item)
                "audio", "music" -> audio.add(item)
                else -> other.add(item)
            }
        }

        val categories = mutableListOf<MediaCategory>()

        if (videos.isNotEmpty()) {
            categories.add(MediaCategory("videos", "Videos", videos))
        }
        if (audio.isNotEmpty()) {
            categories.add(MediaCategory("audio", "Music", audio))
        }
        if (other.isNotEmpty()) {
            categories.add(MediaCategory("other", "Other", other))
        }

        return categories
    }

    // Get stream URL for media item
    fun getStreamUrl(item: MediaItem, quality: String = "auto"): String {
        val encodedPath = URLEncoder.encode(item.path, "UTF-8")
        var url = "$baseUrl/api/stream?path=$encodedPath&quality=$quality"

        authToken?.let {
            url += "&token=$it"
        }

        return url
    }

    // Get thumbnail URL
    fun getThumbnailUrl(item: MediaItem): String? {
        val thumbnail = item.thumbnail ?: return null

        return if (thumbnail.startsWith("http")) {
            thumbnail
        } else {
            "$baseUrl$thumbnail"
        }
    }

    // Search
    suspend fun search(query: String): List<MediaItem> = withContext(Dispatchers.IO) {
        val encodedQuery = URLEncoder.encode(query, "UTF-8")

        val requestBuilder = Request.Builder()
            .url("$baseUrl/api/search?q=$encodedQuery")
            .get()

        authToken?.let {
            requestBuilder.addHeader("Authorization", "Bearer $it")
        }

        val response = client.newCall(requestBuilder.build()).execute()
        val body = response.body?.string() ?: "[]"

        json.decodeFromString(body)
    }

    // Update playback progress
    suspend fun updateProgress(itemId: String, position: Double) = withContext(Dispatchers.IO) {
        val requestBody = """{"id":"$itemId","position":$position}"""
            .toRequestBody("application/json".toMediaType())

        val requestBuilder = Request.Builder()
            .url("$baseUrl/api/progress")
            .post(requestBody)

        authToken?.let {
            requestBuilder.addHeader("Authorization", "Bearer $it")
        }

        client.newCall(requestBuilder.build()).execute()
    }

    // Get playback progress
    suspend fun getProgress(itemId: String): Double = withContext(Dispatchers.IO) {
        val requestBuilder = Request.Builder()
            .url("$baseUrl/api/progress/$itemId")
            .get()

        authToken?.let {
            requestBuilder.addHeader("Authorization", "Bearer $it")
        }

        val response = client.newCall(requestBuilder.build()).execute()
        val body = response.body?.string() ?: """{"position":0}"""

        val progressResponse = json.decodeFromString<ProgressResponse>(body)
        progressResponse.position
    }

    companion object {
        val instance: ApiClient by lazy { ApiClient() }
    }
}

File: appletv/NedflixTV/Sources/APIClient.swift
/*
 * Nedflix for Apple TV - API Client
 * Handles communication with Nedflix server
 */

import Foundation

class APIClient {
    static let shared = APIClient()

    private var baseURL: String = ""
    private var authToken: String?
    private let session: URLSession

    private init() {
        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = 30
        config.timeoutIntervalForResource = 300
        session = URLSession(configuration: config)

        // Load saved settings
        baseURL = UserDefaults.standard.string(forKey: "serverURL") ?? ""
        authToken = UserDefaults.standard.string(forKey: "authToken")
    }

    func configure(serverURL: String) {
        baseURL = serverURL.trimmingCharacters(in: .whitespacesAndNewlines)
        if baseURL.hasSuffix("/") {
            baseURL = String(baseURL.dropLast())
        }
    }

    func setToken(_ token: String?) {
        authToken = token
    }

    // MARK: - Authentication

    func authenticate(username: String, password: String) async throws -> String {
        let url = URL(string: "\(baseURL)/api/auth/login")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")

        let body = ["username": username, "password": password]
        request.httpBody = try JSONEncoder().encode(body)

        let (data, response) = try await session.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse,
              httpResponse.statusCode == 200 else {
            throw APIError.authenticationFailed
        }

        struct AuthResponse: Codable {
            let token: String
        }

        let authResponse = try JSONDecoder().decode(AuthResponse.self, from: data)
        authToken = authResponse.token
        return authResponse.token
    }

    // MARK: - Library

    func fetchLibrary() async throws -> [MediaCategory] {
        let url = URL(string: "\(baseURL)/api/library")!
        var request = URLRequest(url: url)

        if let token = authToken {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }

        let (data, response) = try await session.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse else {
            throw APIError.invalidResponse
        }

        if httpResponse.statusCode == 401 {
            throw APIError.authenticationFailed
        }

        guard httpResponse.statusCode == 200 else {
            throw APIError.serverError(httpResponse.statusCode)
        }

        // Try to decode as array of categories
        do {
            return try JSONDecoder().decode([MediaCategory].self, from: data)
        } catch {
            // Try to decode as flat list and group by type
            do {
                let items = try JSONDecoder().decode([MediaItem].self, from: data)
                return groupItemsByType(items)
            } catch {
                // Return empty if cannot parse
                return []
            }
        }
    }

    private func groupItemsByType(_ items: [MediaItem]) -> [MediaCategory] {
        var videos: [MediaItem] = []
        var audio: [MediaItem] = []
        var other: [MediaItem] = []

        for item in items {
            switch item.type.lowercased() {
            case "video", "movie", "tv":
                videos.append(item)
            case "audio", "music":
                audio.append(item)
            default:
                other.append(item)
            }
        }

        var categories: [MediaCategory] = []

        if !videos.isEmpty {
            categories.append(MediaCategory(id: "videos", name: "Videos", items: videos))
        }
        if !audio.isEmpty {
            categories.append(MediaCategory(id: "audio", name: "Music", items: audio))
        }
        if !other.isEmpty {
            categories.append(MediaCategory(id: "other", name: "Other", items: other))
        }

        return categories
    }

    // MARK: - Streaming

    func getStreamURL(for item: MediaItem, quality: String = "auto") -> URL? {
        var urlString = "\(baseURL)/api/stream?path=\(item.path.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? item.path)"

        urlString += "&quality=\(quality)"

        if let token = authToken {
            urlString += "&token=\(token)"
        }

        return URL(string: urlString)
    }

    func getThumbnailURL(for item: MediaItem) -> URL? {
        if let thumbnail = item.thumbnail, !thumbnail.isEmpty {
            if thumbnail.hasPrefix("http") {
                return URL(string: thumbnail)
            } else {
                return URL(string: "\(baseURL)\(thumbnail)")
            }
        }
        return nil
    }

    // MARK: - Search

    func search(query: String) async throws -> [MediaItem] {
        let encoded = query.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? query
        let url = URL(string: "\(baseURL)/api/search?q=\(encoded)")!

        var request = URLRequest(url: url)
        if let token = authToken {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }

        let (data, _) = try await session.data(for: request)
        return try JSONDecoder().decode([MediaItem].self, from: data)
    }

    // MARK: - Playback State

    func updateProgress(itemId: String, position: Double) async throws {
        let url = URL(string: "\(baseURL)/api/progress")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")

        if let token = authToken {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }

        let body: [String: Any] = [
            "id": itemId,
            "position": position
        ]
        request.httpBody = try JSONSerialization.data(withJSONObject: body)

        _ = try await session.data(for: request)
    }

    func getProgress(itemId: String) async throws -> Double {
        let url = URL(string: "\(baseURL)/api/progress/\(itemId)")!
        var request = URLRequest(url: url)

        if let token = authToken {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }

        let (data, _) = try await session.data(for: request)

        struct ProgressResponse: Codable {
            let position: Double
        }

        let response = try JSONDecoder().decode(ProgressResponse.self, from: data)
        return response.position
    }
}

// MARK: - Errors

enum APIError: LocalizedError {
    case invalidURL
    case invalidResponse
    case authenticationFailed
    case serverError(Int)
    case networkError(Error)

    var errorDescription: String? {
        switch self {
        case .invalidURL:
            return "Invalid server URL"
        case .invalidResponse:
            return "Invalid server response"
        case .authenticationFailed:
            return "Authentication failed"
        case .serverError(let code):
            return "Server error: \(code)"
        case .networkError(let error):
            return error.localizedDescription
        }
    }
}
